### PR TITLE
fix: improve path collision detection and simplify check

### LINKED ISSUE #890: Error: Duplicate path parameter definition signature found in controller...
The new version throws an error, which I can not exactly figure out:
Generate routes error.
` Error: Duplicate path parameter definition signature found in controller SitemapController at [method GET getJobsSitemap, getLpSitemap, getLpSitemapIndex], JobfeedController at [method GET getJobFeedsPrioBranch, getJobFeedsPrioDefault, getJobFeedsBranchRegion, getJobFeedsBranch, getJobFeedsDefault]`

We have methods like this:
```ts
@Get('/jobprio-{page}-{prio}-{branch}.{format}')
	public async getJobFeedsPrioBranch(
		@Request() req: express.Request,
		@Path('page') page: string,
		@Path('format') format: JobfeedOutputs, // 'xml' | 'csv' | 'json',
		@Path('prio') prio?: JobPrioLevel | '0',
		@Path('branch') branch?: string,
		@Query('utm_campaign') utm_campaign?: string,
		@Query('utm_source') utm_source?: string,
		@Query('utm_medium') utm_medium?: string
	)
```

```ts
@Get('/jobprio-{page}-{prio}.{format}')
	public async getJobFeedsPrioDefault(
		@Request() req: express.Request,
		@Path('page') page: string,
		@Path('format') format: JobfeedOutputs, // 'xml' | 'csv' | 'json',
		@Path('prio') prio?: JobPrioLevel | '0',
		@Query('utm_campaign') utm_campaign?: string,
		@Query('utm_source') utm_source?: string,
		@Query('utm_medium') utm_medium?: string
	)
```

this worked before, and should still work right?

## Sorting

- **I'm submitting a ...**

  - [x] bug report
  - [ ] feature request
  - [ ] support request

- I confirm that I
  - [x] used the [search](https://github.com/lukeautry/tsoa/search?type=Issues) to make sure that a similar issue hasn't already been submit

## Expected Behavior
the routes should be computed correctly.

## Current Behavior

throws error

## Possible Solution

not quite sure yet where the error is coming from..maybe from the path parameters?

## Steps to Reproduce

add two methods with different dynamic path paramters. (see example above)

## Context (Environment)

Version of the library: 3.5.0
Version of NodeJS: 15.2.1

### PR DESCRIPTION
**Pull request by simllll**
### All Submissions:

- [x] Have you followed the guidelines in our [Contributing](https://github.com/lukeautry/tsoa/tree/master/docs/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/lukeautry/tsoa/pulls) for the same update/change?
- [x] Have you written unit tests?
- [x] Have you written unit tests that cover the negative cases (i.e.: if bad data is submitted, does the library respond properly)?
- [x] This PR is associated with an existing issue?

closes #890 #894

**Test plan**

I added 2 more cases to the route tests:
* two parameters in one route without a folder
* two routes but with a different "prefix"

### CODE REVIEW

File: packages/cli/src/metadataGeneration/metadataGenerator.ts [hunk 1]
```diff
@@ -113,7 +113,7 @@ export class MetadataGenerator {
   };
 
   private checkForPathParamSignatureDuplicates = (controllers: Tsoa.Controller[]) => {
-    const isParamRegExp = new RegExp('{|}|:');
+    const paramRegExp = new RegExp('{(\\w*)}|:(\\w+)', 'g');
```
Comment (simllll) [inline packages/cli/src/metadataGeneration/metadataGenerator.ts:116]: is there any regex pre defined for parsing the parameters?

File: packages/cli/src/metadataGeneration/metadataGenerator.ts [hunk 2]
```diff
@@ -125,16 +125,12 @@ export class MetadataGenerator {
...
+      subject: { paths: Array<{ cntParams: number; path: string }>; method: Tsoa.Method },
     ): PathDuplicationType {
       const testLength = tester.paths.length > subject.paths.length ? subject.paths.length : tester.paths.length;
       for (let i = 0; i < testLength; i += 1) {
-        if (
-          (tester.paths[i].isParam && !subject.paths[i].isParam) ||
-          (!tester.paths[i].isParam && subject.paths[i].isParam) ||
-          (!tester.paths[i].isParam && tester.paths[i].path !== subject.paths[i].path)
-        ) {
+        if (tester.paths[i].cntParams !== subject.paths[i].cntParams || tester.paths[i].path !== subject.paths[i].path) {
```
Comment (simllll) [inline packages/cli/src/metadataGeneration/metadataGenerator.ts:133]: just came into my mind: I think we could probably even more simplify this check, cntParms is not exactly needed anymore, as I compare the path with empty placeholders now.

Comment (simllll) [reply packages/cli/src/metadataGeneration/metadataGenerator.ts]: simplified it

File: packages/cli/src/metadataGeneration/metadataGenerator.ts (hunk 1 above)
Comment (WoH) [reply packages/cli/src/metadataGeneration/metadataGenerator.ts:116]: Ideally, for express/koa: Replace `{}` with `:` and feed that into [path-to-regexp](https://www.npmjs.com/package/path-to-regexp)

Comment (simllll) [reply packages/cli/src/metadataGeneration/metadataGenerator.ts:116]: I understand, I will push a new rewrite with simplifies and improves the check even more.. and kicks out "false flags" ;)

**Comment by simllll** [comment]
I refactored the code, simplified it and improved the collision detection... it should not trigger any "fake" warnings any more, but it warns in two cases now:
1.) if the path is exactly the same (regardless if they contain a parameter or not, but probably most common with parameters ;-))
2.) if the path collides in order, this check only applies to routes with parameters: e.g. 
a. /route-:param1
b. /route-:param2-:param3
Here route b collides with route a. As "param1" could be "a-b".

This one will not collide though:
a. /route-:param1
b. /route-:param2/:param3
as the path is seperated by a "/".

I also changed the output of the warnings.

**Comment by simllll** [comment]
This issue (and the unreleased issue #896) is quite a show stopper for our team right now. Anything I can do to bring this into master and publish a new release? :) thanks a lot for all your effort @WoH

**Comment by WoH** [comment]
> This issue (and the unreleased issue #896) is quite a show stopper for our team right now. Anything I can do to help bringing this into master and a get new tsoa release? :) thanks a lot for all your effort @WoH 

I'm sorry about that. The code looked good, but I wanted to play around with it for myself to prevent any avoidable issues. LGTM, thanks a lot for the PR!

**Pull request merged**